### ISSUE TITLE
Context managers in `Depends` are broken after 0.106

### ISSUE THREAD

**Issue by Kludex** [tiangolo/fastapi#11143]
### Discussed in https://github.com/tiangolo/fastapi/discussions/11107

<div type='discussions-op-text'>

<sup>Originally posted by **FeeeeK** February  7, 2024</sup>
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Request

app = FastAPI()


class Session:
    def __init__(self):
        print("creating session")

    async def __aenter__(self):
        print("opening session")
        return self

    async def __aexit__(self, exc_type, exc, tb):
        print("closing session")

    async def commit(self):
        print("committing session")

    async def rollback(self):
        print("rolling back session")


@app.middleware("http")
async def commit_session(request: Request, call_next):
    # minimalistic middleware for example, my code uses ASGI middleware
    response = await call_next(request)
    db_session = request.scope.get("db_session")
    if not db_session:
        return response

    if response.status_code // 200 != 1:
        await db_session.rollback()
    else:
        await db_session.commit()

    return response


async def get_db_session(request: Request):
    async with Session() as session:
        request.scope["db_session"] = session
        yield session


@app.get("/")
async def root(session: Session = Depends(get_db_session)):
    return {"message": "Hello World"}


# Pre 0.106 behaviour:

# creating session
# opening session
# committing session
# closing session

# Post 0.106 behaviour:
# creating session
# opening session
# closing session
# committing session

# The session is not committed, because it's closed before the middleware is called.
```

### Description

Before `0.106`, Depends execution after yield was after middlewares, which allowed to access resources created for a route (e.g. sessions) and do something with them depending on the response (which cannot be done with Depends), but after 0.106, the behavior has changed and this feature is no longer available. The documentation only talks about background tasks, but not a word about middlewares. Was this behavior change intentional?

### Operating System

Windows

### FastAPI Version

0.109.2

### Pydantic Version

2.4.2

### Python Version

3.11.1

### Additional Context

_No response_</div>

**Comment by falkben** [comment]
I [commented](https://github.com/tiangolo/fastapi/discussions/11107#discussioncomment-8459459) in the Discussion that `StreamingResponse`'s no longer work with `Depends` resources. Since this is a pretty large change in behavior, and wasn't called out in the release notes, I wonder if this is intended. Should this be a separate Issue?

**Comment by Kludex** [comment]
> I [commented](https://github.com/tiangolo/fastapi/discussions/11107#discussioncomment-8459459) in the Discussion that `StreamingResponse`'s no longer work with `Depends` resources. Since this is a pretty large change in behavior, and wasn't called out in the release notes, I wonder if this is intended. Should this be a separate Issue?

No. The only reason I created this issue because of your comment. 🙏

**Comment by Kludex** [comment]
Does this comment by @tiangolo answers this issue: https://github.com/tiangolo/fastapi/discussions/11177#discussioncomment-8559300 ?

**Comment by falkben** [comment]
I might be misinterpreting but I don't think so? Seems like it was intended to disable using resources in background tasks. But was it intended to also disable using resources in path operators?

**Comment by mortalisk** [comment]
It seems this is what breaks our streaming endpoints, the lack of which kills throughput making everything slow.

**Comment by mortalisk** [comment]
A workaround for now is to use our connection pool directly in each endpoint doing streaming. Annoying, but works.
Doing something like this for a psycopg pool:
```python
@router.get("/stuff", response_model=list[Stuff])
async def get_all_stuff(pool: PoolDep):
    connection = await pool.getconn()

    async def commit_and_put_in_pool():
        await connection.commit()
        await pool.putconn(connection)

    return StreamingResponse(
            stream_stuff(connection), media_type="application/json", background=commit_and_put_in_pool
    )
```
This seems to work for us. But is it correct? Will the background task always run after the streaming is done? Or can we end up in a situation where it is run in the middle of streaming?

**Comment by brian-goo** [comment]
Out streaming endpoints are broken probably due to this.
So it seems that there is no `good` workaround as long as a db session is managed with `Depends`?

**Comment by scriptator** [comment]
The release notes suggest the following:

> If you used to rely on this behavior, now you should create the resources for background tasks inside the background task itself, and use internally only data that doesn't depend on the resources of dependencies with yield.
> 
> For example, instead of using the same database session, you would create a new database session inside of the background task, and you would obtain the objects from the database using this new session. And then instead of passing the object from the database as a parameter to the background task function, you would pass the ID of that object and then obtain the object again inside the background task function.

i.e. just don't use `Depends` for obtaining your db session for background tasks any more but obtain it within the background task

**Comment by falkben** [comment]
@scriptator we're not talking about background tasks here.

**Comment by scriptator** [comment]
> @scriptator we're not talking about background tasks here.

I was replying to @brian-goo who talked about streaming responses which behave like background tasks in this regard. I had a problem because I used a db session in a streaming response that I obtained with Depends and it was fixed by following the advice for background tasks: initiate your resources within the background task (or generator function in the case of streaming responses).

But you are right if you meant that this does not contribute to the original problem from the issue.

**Comment by falkben** [comment]
I don't think streaming responses should be considered similar to background tasks. They are part of the request/response cycle. I think their behavior in this case being similar to background tasks not being able to use dependency injection in the path operators is likely a bug.

**Comment by mortalisk** [comment]
Any progress on this. Our workaround is not working as expected. It seems the background task is not guaranteed to run, and thus our connection pool is failing when ever there is any error, essentially taking down the entire program, and we must restart.

This basically renders streaming useless for us.

Is there something else we can do to make it work?

**Comment by hasansustcse13** [comment]
I am facing the same problem. Does anybody know any workaround for this problem.
#11444

**Comment by falkben** [comment]
For now, just staying on version `<0.106`.

**Comment by mortalisk** [comment]
Does `<0.106` work on python 3.12? Because 0.109 says "Add support for Python 3.12."

**Comment by btemplep** [comment]
+1 for this issue.  It breaks request context management for stream proxies.  

ie  I have a `Depends` parameter for connecting to an external streaming API.  I start a stream from an external API in the route.  I need the initial response from the external API to contribute to the `StreamResponse` return.  Then  I need that same connection to stream from my streaming generator.   

The work around in my case is to use a global client for the external streaming API so that cleanup is not called on the return of `StreamingResponse`.   As well as managing the context with try/finally blocks in the route and streaming generator.   Then more context managers in the streaming generator for other `Depends` that die when the route returns.

**Comment by libe** [comment]
@btemplep could you provide a code example for your workaround?

**Comment by fangpenlin** [comment]
We also encountered issues with streaming response after upgrading FastAPI to > 0.106.

We implemented Git service in HTTP protocol. During the time streaming response is generating content to the client, it needs to access to database session and other services provided bye `Depends`. While the rationale of the change is understandable, but streaming response is very different from background task. As the moment we are streaming the response, the request is still **alive**. The streaming content could still dynamically changes based on the underlying resources or connections to database or other services. So at the moment of streaming content, we still need to access the dependencies. Because if we have all the response content ready when returning the route function, we won't need to use streaming response anyway, as it defeats its own purpose.

Is it possible to delay the destruction of dependencies only after the HTTP session actually closed (no more content will be sent to the client)? Or make it a special case if the returned response is a streaming response?

**Comment by btemplep** [comment]
@libe Here is an example of what I expected to work

```python 
from typing import Any, AsyncGenerator, Annotated

import aioboto3
from fastapi import Depends, FastAPI
from fastapi.responses import StreamingResponse

from sqlalchemy.ext.asyncio import AsyncAttrs, async_sessionmaker, AsyncSession, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

engine = create_async_engine("sqlite+aiosqlite:///test.sqlite")
sessionmaker: async_sessionmaker[AsyncSession] = async_sessionmaker(expire_on_commit=False)


class Base(AsyncAttrs, DeclarativeBase):
    pass


class SomeTable(Base):
    __tablename__ = "some_table"

    some_id: Mapped[int] = mapped_column(primary_key=True, nullable=False)


async def get_db() -> AsyncGenerator[AsyncSession, None, None]:
    async with sessionmaker() as sess:
        yield sess


async def get_boto_client():
    async with aioboto3.Session().client("bedrock-runtime") as br_client:
        yield br_client


app = FastAPI()


async def _stream_resp(db_sess: AsyncSession, stream_resp: dict):
    async for chunk in stream_resp['body']:
        # also do something with the db_session
        yield chunk


@app.route("/route_here")
async def route_here(
    db_sess: Annotated[AsyncSession, Depends(get_db)],
    br_client: Annotated[Any, Depends(get_boto_client)]
):
    stream_resp = br_client.invoke_model_with_response_stream(args="here")
    return StreamingResponse(
        _stream_resp(
            db_sess=db_sess,
            stream_resp=stream_resp
        )
    )
````

Here is a simplified version of what my fix is. 

```python
from contextlib import asynccontextmanager
from typing import Any, AsyncGenerator, Annotated

import aioboto3
from fastapi import Depends, FastAPI
from fastapi.responses import StreamingResponse

from sqlalchemy.ext.asyncio import AsyncAttrs, async_sessionmaker, AsyncSession, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

engine = create_async_engine("sqlite+aiosqlite:///test.sqlite")
sessionmaker: async_sessionmaker[AsyncSession] = async_sessionmaker(expire_on_commit=False)
br_client = None


class Base(AsyncAttrs, DeclarativeBase):
    pass


class SomeTable(Base):
    __tablename__ = "some_table"

    some_id: Mapped[int] = mapped_column(primary_key=True, nullable=False)


async def get_db() -> AsyncGenerator[AsyncSession, None, None]:
    async with sessionmaker() as sess:
        yield sess


@asynccontextmanager
async def lifespan(app: FastAPI):
    global br_client
    async with aioboto3.Session().client("bedrock-runtime") as br:
        br_client = br
        yield


app = FastAPI(
    lifespan=lifespan
)


async def _stream_resp(stream_resp: dict):
    # In the streaming generator we need to start our own context manager for DB
    async with sessionmaker() as db_sess:
        # Because the client is global the stream on the backend is preserved. 
        async for chunk in stream_resp['body']:
            # also do something with the db_session
            yield chunk


@app.route("/route_here")
async def route_here(
    db_sess: Annotated[AsyncSession, Depends(get_db)]
):
    # br_client is now global, and shared
    stream_resp = br_client.invoke_model_with_response_stream(args="here")
    # do DB stuff here, we can still use the Depends
    return StreamingResponse(
        _stream_resp(
            db_sess=db_sess,
            stream_resp=stream_resp
        )
    )
````

**Comment by RamiAwar** [comment]
I agree, this is definitely unexpected behavior. I understand the motivation for not wanting to hold resources like db sessions more than needed, but it does make sense in the context of a streaming response.

If we're yielding from a generator and we want to commit when the generator is fully consumed (as part of the generator, calling session.commit() after the generator stop iteration is raised and the response is closed) then it makes sense to keep the session resource available!

**Comment by buptzyf** [comment]
@tiangolo hello，do you have any ideas？

**Comment by tosmart01** [comment]
I encountered an issue with SQLAlchemy sessions not closing properly when using StreamingResponse in FastAPI. The problem was that the session was not being closed, which led to warnings about unclosed connections.

I found that explicitly closing the session by using await session.close() in the generator function resolved the issue. Here’s how I implemented it:

async def get_user(db):
    await db.get(User, User.id==1)
    await db.close()


@app.route("/user-info")
async def user(db: Depends(get_db)):
    return StreamingResponse(
        get_user(db)
    )

I hope this helps others encountering a similar issue!

Best regards,

**Comment by RamiAwar** [comment]
> I think Depends should accept a flag where you can specify when the dependency should be closed, before or after the response.
> 
> Closing before is useful if you want to raise HTTPExceptions from the cleanup step of your dependencies. Not something I've ever done but maybe if I was just getting started with FastAPI and and learning about that feature, I would find it useful.
> 
> Closing after is useful with streamsresponses, if you need the dependencies in background tasks, or when you want to perform slowish operations in the cleanup to be able to send the response faster.

and since this was a breaking change, I'd make it "after" by default for backwards compatibility.

I'm personally still locked to .105 and missing security updates cause of this.

**Comment by sherpya** [comment]
what about?
`return StreamingCSVResponse(csv_stream, headers=headers, background=BackgroundTask(session.close))`
it worked for me but I'm not sure it's the correct way to go

**Comment by Kludex** [comment]
> Is there any update on this @tiangolo @Kludex ? I'm trying to figure out if we need to rewrite our `Depends` resource acquisition functions to be compatible with StreamingResponses.

I've pinged Sebastián. That's what I can do.

**Comment by tiangolo** [comment]
Thanks @falkben! And thanks @Kludex for the ping and help. :bow: 

Indeed, agreed that this is an unintended bug. And streaming responses are important for me/FastAPI, in fact I want to improve internal support for them, I have to experiment with that.

I haven't sat down to figure out the right way to solve this yet, but yes, it's under my radar.

The original intention with having dependencies close after the request is done was to allow using them in background tasks, thinking they could use the same database session. But later I realized that the work in a background task could end up being done much later, and keeping the same session would mean holding a DB connection that was not really needed for a long time (waiting for the network when the response is sent, so that's why I made dependencies close before sending the response. But I didn't realize that streaming responses would return something from the function, but would need the resource (DB session/connection) available until the response is finished sending, that was an oversight from my part.

I'm not sure what's the right way to fix it, I've considered a few approaches: having some attribute in responses marking them as finished in some way (the data is already there, they don't need any extra resource), so a streaming response would not be finished. Another alternative would be holding dependencies available and open when the return is a `Response` object, and only close them early if the response is an arbitrary object that FastAPI would automatically serialize.

**Comment by falkben** [comment]
Thanks very much! Could an extra attribute in the Depends call indicate whether to keep it around?

I also like the idea of keeping the dependency available if the return is a Response object, and hopefully wouldn't add much overhead.

**Comment by truh** [comment]
I think it would still be beneficial to be able to use the dependencies in the background tasks.
* To reduce computation for creating and closings decencies.
* To avoid duplicate code for creating more complex dependencies.
* The dependency might contain state from the request handler that can be used in the background task.

This doesn't necessarily mean that all dependencies need to kept open until after the request is done. If there was a way to declare which dependencies the background task needs, the other dependencies could be closed early.

**Comment by schneebuzz** [comment]
This also breaks things like this without `Depends`:
```python
@app.route("/get-file")
async def get_file():
    async with get_some_file_stream() as some_async_iterable:
        return StreamingResponse(
            content=some_async_iterable,
            headers=...,
            media_type=...,
        )
```

as the context is exited before finishing response streaming. I think this is not the expected behavior.

**Comment by truh** [comment]
@schneebuzz this seems expected to me. Exiting the context manager closes the context. StreamResponse finishes only after it's returned.

If depends only closed after the response has completed, you could use them to turn the context into a procedure that returns the iterable and is cleaned up after the request has completed.

**Comment by vtgn** [comment]
Hi!
I noticed the same problem (version 0.115.4), and I was confused for a moment because the documentation says the contrary: https://fastapi.tiangolo.com/tutorial/middleware/?h=middleware

> If you have dependencies with yield, the exit code will run after the middleware.

**You should at least modify this wrong documentation, and give more information about this until the final specification has been chosen and implemented to avoid to mislead and confuse the developers.**

In my case, I would like to commit the DB session only when it is certain that the 2xx status response will be sent without any problem. But during the middleware execution, unexpected problems may happen and the sent response may be in error status, which is annoying because the client thinks his request failed while it was stored inside the DB. So closing the session at the end of the middleware would have been a good solution, so this is a shame that this feature doesn't work anymore :/.

The only workaround I found is to not use the `yield` feature, but to create, close and manage the commit/rollback of the session inside the middleware, and to store its instance into `Request.state` object to share it between the middleware and the endpoint, then a `starlette_context` (https://starlette-context.readthedocs.io/en/latest/fastapi.html) allows to inject the session in any method/function without passing the session as argument. FYI the `starlette_context` is not available from the middleware method, like the `Depends` of FastApi.

Example:
```py
from fastapi import APIRouter, Depends, FastAPI, Request
from sqlalchemy.orm import Session
from starlette_context import context, request_cycle_context

# The starlette context feature, init a new context for each request/response cycle:
async def my_context_dependency():
    data = {}
    with request_cycle_context(data):
        yield

app = FastAPI(dependencies=[Depends(my_context_dependency)])


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):

    session: Final[Session] = db.createSession() # method of your own to create a new Session instance, no yield inside.
    request.state.session = session # store the session inside the request to share it with the endpoint method.

    request.state.any_data = ... # store anything you want, it will be available by the method of the endpoint.

    try:
        response = await call_next(request) # call the endpoint method.

        any_other_data = request.state.any_other_data # Exists because it has been set in the endpoint method.

        session.commit()
    except Exception as e:
        session.rollback()
        raise e
    finally:
        session.close()

    return response

# Optional method, but useful if you want to add the session inside the Fastapi injectable dependencies.
def get_session(request: Request) -> Session:
    return request.state.session # Exists because it has been set by the middleware method.

# Task launched at the end of the endpoint by using the BackgroundTasks feature (see below), after having sent the response to the client (the last response for streaming response)
def end_task(any_data, any_other_data, ...):
    ...

# The endpoint method:
app.post("")
def endpoint(
    request: Request,
    background_task: BackgroundTasks,
    session: Session = Depends(get_session) # Optional. If you don't use this injection, the session is inside request.state.session set by the middleware.
):
    context["session"] = session # Store the session inside the starlette context.

    response = db_service() # Defined in another module (see below), no need to pass the session as argument.
    
    any_data = request.state.any_data # exists because it has been set in the middleware method.

    request.state.any_other_data = ... # store inside `request.state` anything you want, it will be available in the middleware when returning to it on response.

    background_task.add_task(end_task, any_data, any_other_data,...) # you can pass anything you want

    return response


### In another Python module for the db services
from sqlalchemy.orm import Session
from starlette_context import context

def db_service(session: Session = None): # impossible to pass the context as default value, it generates an error during the import of the module, because it exists only during a request/response cycle.
    if session is None: # if no session is given in arg, use the one of the context of the current endpoint
        context["session"]

    # Any commands in DB
    ...

    return result

```

I think you can find a similar solution for your problems of stream.

**To summarize:**
- you don't need to use the `yield` feature.
- share data between the middleware and the endpoint by the `request.state`, it works both ways (ie. you can also set a new field in `request.state` from the endpoint method, it will be available in the middleware when returning to it on response.
- use starlette context to retrieve anything you want from anywhere during a request/response cycle.
- use BackgroundTasks to pass and launch anything you want after the sending of the response (the sending of the last response for streaming response) 

For example, I could have managed the commit/rollback/closing of the db session from a background task if I want to continue to use the session after the sending of the response, or if only some endpoints needs a db session, I can create a session in the endpoint method and send it to the middleware or to a background task as explained.

Regards.

**Comment by johnboxall** [comment]
I got bit by this when using `psycopg.AsyncConnection` with `StreamingResponse`.

I want to use a pattern like described in this [article](https://blog.danielclayton.co.uk/posts/database-connections-with-fastapi/#add-on-using-a-psycopg-connection-pool) where my connections is declared using `Depends`:

```python
import os
from typing import Annotated

import psycopg
import uvicorn
from fastapi import Depends, FastAPI
from fastapi.responses import StreamingResponse

app = FastAPI()


async def get_conn():
    async with await psycopg.AsyncConnection.connect(
        "postgresql://user:pass@localhost/db"
    ) as conn:
        yield conn


conn_dep = Annotated[psycopg.AsyncConnection, Depends(get_conn)]


@app.get("/")
async def stream(conn: conn_dep):
    async with conn.cursor() as cursor:

        async def _stream():
            await cursor.execute("SELECT 42")
            row = await cursor.fetchone()
            yield f"data: {row}\n\n"

        return StreamingResponse(_stream(), media_type="text/event-stream")


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=int(os.getenv("PORT", "8000")), reload=True)
```

But, I get the following error:

```
Traceback (most recent call last):
  File "starlette/responses.py", line 264, in wrap
    await func()
  File "starlette/responses.py", line 245, in stream_response
    async for chunk in self.body_iterator:
  File "test.py", line 27, in _stream
    await cursor.execute("SELECT 42")
  File "psycopg/cursor_async.py", line 97, in execute
    raise ex.with_traceback(None)
psycopg.OperationalError: the connection is closed
```

Since I want to continue to use the latest version of FastAPI I restructured my code to manually manage the connection:

```python
import os

import psycopg
import uvicorn
from fastapi import FastAPI
from fastapi.responses import StreamingResponse

app = FastAPI()


@app.get("/")
async def stream():
    conn = await psycopg.AsyncConnection.connect("postgresql://user:pass@localhost/db")
    cursor = conn.cursor()

    async def _stream():
        try:
            await cursor.execute("SELECT 42")
            row = await cursor.fetchone()
            yield f"data: {row}\n\n"
        finally:
            await cursor.close()
            await conn.close()

    return StreamingResponse(_stream(), media_type="text/event-stream")


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=int(os.getenv("PORT", "8000")), reload=True)
```

**Comment by aschreiber1** [comment]
Is there any update on this, we've been stuck on version 0.105 for 18 months at this point...

**Comment by aschreiber1** [comment]
@tiangolo  Just trying to understand the thinking on this one. This seems like Fast API is dying in a sense that no progress is being made on a pretty fundamental broken part of the system, should users just take this as this will never work and we should find alternatives to Fast API?

**Comment by tsangwpx** [comment]
lifetime topic is as hard as in rust.

My two cents here:
- One way to solve the issue is to consume (await) the response object in get_request_handler(), but the background task may occupy the context and resources indefinitely.
- Other way is to inspect the ASGI message, and manage the async_exit_stack manually, so that the stack is closed after `more_body=False`.

Both need big refactor, and probably cause incompatibilities.

Lastly, if the dependency chain is simple, we may inject the database connection factory into the streaming response. and create a new database object there. (of course, this assume the lifetime of the factory (and its dependencies) outlive the request handler, and this is usually true because the factory is usable during process lifetime).

```python
engine = create_engine(DB_URI)
Session = sessionmaker(bind=engine)

def get_db_factory() -> Callable[[], Session]:
   return Session

def iterator(
    db_factory,
) -> Iterator[str]:
   db = db_factory()
   with db:
     yield 'hello'
     db.commit()
     yield 'world'
     db.commit()

@router.get("/stream")
def hello(  
    db_factory = Depends(get_db_factory),
) -> StreamingResponse:
    return StreamingResponse(iterator(db_factory))
```

**Comment by tiangolo** [comment]
Support for `StreamingResponse`s using dependencies with `yield` should be fixed by this: https://github.com/fastapi/fastapi/pull/14099 (once it's merged and released). 🚀

**Comment by aschreiber1** [comment]
Amazing!

**Comment by tiangolo** [comment]
FastAPI 0.118.0 was just released with the fix in https://github.com/fastapi/fastapi/pull/14099. 🐛 🚀 🎉
